Add pull requests component

iam-flo opened the pull request:
### Motivation
Adds a pull request widget component and the necessary data to display pull requests of a user

### Description
- extended Pull request model and api
- add widget for pull request

### Screenshots (if applicable)
<!-- Attach screenshots here. -->
<img width="294" alt="Bildschirmfoto 2024-09-16 um 18 06 58" src="https://github.com/user-attachments/assets/3a0c2719-da10-4092-a13b-fe97b09a700a">

### Checklist

#### General

- [x] PR title is clear and descriptive
- [x] PR description explains the purpose and changes
- [x] Code follows project coding standards
- [x] Self-review of the code has been done
- [x] Changes have been tested locally
- [x] Screenshots have been attached (if applicable)
- [x] Documentation has been updated (if applicable)

#### Client (if applicable)

- [x] UI changes look good on all screen sizes and browsers
- [x] No console errors or warnings
- [x] User experience and accessibility have been tested
- [x] Added Storybook stories for new components
- [ ] Components follow design system guidelines (if applicable)

#### Server (if applicable)

- [x] Code is performant and follows best practices
- [x] No security vulnerabilities introduced
- [x] Proper error handling has been implemented
- [x] Added tests for new functionality
- [x] Changes have been tested in different environments (if applicable)

FelixTJDietrich commented on webapp/src/app/core/modules/openapi/model/pull-request-label.ts:
```diff
@@ -0,0 +1,18 @@
+
+
+export interface PullRequestLabel { 
+    name?: string;
```
Is there a way to make it not optional? I think if it exists it is always defined with `name` and `color`

GODrums replied:
This seems to be a general issue with JPA as OpenAPI Java-Variables can always be null implicitly. It applies to all our other model interfaces too.

FelixTJDietrich commented on server/application-server/src/main/java/de/tum/in/www1/hephaestus/codereview/pullrequest/PullRequestLabel.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Embeddable
```
So this means that there is no separate table for it am I right?

iam-flo replied:
yes, but we could also create an own table for labels if we think that it is necessary. but then we have to handle the labels more

FelixTJDietrich commented on webapp/src/app/ui/pull-request-widget/pull-request-widget.component.html:
```diff
@@ -0,0 +1,21 @@
+  <div class="flex justify-between items-center mb-2 text-xs text-gray-500">
+    <span class="font-bold">{{ pullRequest().repository.name }} #{{ pullRequest().number }} on {{ pullRequest().createdAt }}</span>
+    <span class="flex items-center space-x-2">
+      <span class="text-green-600 font-bold">+{{ pullRequest().additions }}</span>
```
Here you should be using the GitHub color, inspect said they are using `color-fg-success` this maps to `text-github-success-foreground` on our side I believe.

FelixTJDietrich commented on webapp/src/app/ui/pull-request-widget/pull-request-widget.component.html:
```diff
@@ -0,0 +1,21 @@
+    <span class="font-bold">{{ pullRequest().repository.name }} #{{ pullRequest().number }} on {{ pullRequest().createdAt }}</span>
+    <span class="flex items-center space-x-2">
+      <span class="text-green-600 font-bold">+{{ pullRequest().additions }}</span>
+      <span class="text-red-600 font-bold">-{{ pullRequest().deletions }}</span>
```
Use github color they use `danger` here

FelixTJDietrich commented on webapp/src/app/ui/pull-request-widget/pull-request-widget.component.html:
```diff
@@ -0,0 +1,21 @@
+<div class="border border-gray-300 rounded-lg p-4 w-72">
+  <div class="flex justify-between items-center mb-2 text-xs text-gray-500">
+    <span class="font-bold">{{ pullRequest().repository.name }} #{{ pullRequest().number }} on {{ pullRequest().createdAt }}</span>
```
Can you als add an artifact icon, e.g.:
https://primer.style/components/icon

Maybe also name it more generic, this could be an entity that we can also use for issues. 

You can reference the leaderboard how to include an icon from octicons

FelixTJDietrich commented on webapp/src/app/ui/pull-request-widget/pull-request-widget.component.html:
```diff
@@ -0,0 +1,21 @@
+  <div class="flex justify-between font-bold text-sm mb-3">
+    {{ pullRequest().title }}
+    @if (pullRequest().reviews) {
+      <span class="text-xs text-gray-500">({{ pullRequest().reviews.size }} reviews)</span>
```
Here you can just show number of approvals with an check icon, change requests, and comments.

FelixTJDietrich commented on webapp/src/app/ui/pull-request-widget/pull-request-widget.component.ts:
```diff
@@ -0,0 +1,11 @@
+import { PullRequest } from '@app/core/modules/openapi';
+
+@Component({
+  selector: 'app-pull-request-widget',
```
I don't like widget, we could call it `app-issue-card` because it can also be used for issues in the future and pull requests are a subclass of issues

FelixTJDietrich commented on webapp/src/app/ui/pull-request-widget/pull-request-widget.stories.ts:
```diff
@@ -0,0 +1,50 @@
+import { PullRequest, PullRequestReview, Repository } from '@app/core/modules/openapi';
+
+const meta: Meta<PullRequestWidgetComponent> = {
+  title: 'Components/PullRequestCard',
```
Naming is inconsistent

FelixTJDietrich commented:
We also have colors defined for border and card that you can make use off

GODrums commented on server/application-server/src/main/java/de/tum/in/www1/hephaestus/codereview/pullrequest/PullRequest.java:
```diff
@@ -36,6 +30,9 @@ public class PullRequest extends BaseGitServiceEntity {
     private String url;
 
+    @NonNull
+    private int prNumber;
```
I already included the `number`-variable this in another PR . When you bring this branch up-to-date with `develop` you can just use that one instead.

GODrums commented on webapp/src/app/ui/app-issue-card/app-issue-card.component.html:
```diff
@@ -0,0 +1,31 @@
+<div class="border border-gray-300 rounded-lg p-4 w-72">
+  <div class="flex justify-between items-center mb-2 text-xs text-gray-500">
+    <span class="font-bold flex justify-center items-center space-x-1">
+      <ng-icon [svg]="octGitPullRequest" size="16" class="mr-1"></ng-icon> {{ pullRequest().repository.name }} #{{ pullRequest().number }} on {{ pullRequest().createdAt }}
```
You have to handle the `null`-case for the repository here to avoid a compiler error.

I'd be fine with simple optional chaining:
`{{ pullRequest().repository?.name }}`

GODrums reviewed (changes requested):
I wasn't able to run storybook in the currently version due to a naming conflict of the `number`/`prNumber`-variable.

I would recommend that you rebase the branch to `develop` as it already contains the `number`-attribute.

iam-flo commented:
Rework and ready to merge

FelixTJDietrich commented:
It looks like this in the Storybook now:
![image](https://github.com/user-attachments/assets/6110ba1f-1905-4356-93dc-a5503056c621)
https://66a8981a27ced8fef3190d41-wvdgwzpzof.chromatic.com/?path=/docs/ui-appissuecard--docs

Some points that I noticed, see screenshot:
- You can remove the `App` prefix, this is only important for the selector
- The `IssueCard` is a component building on the blocks in the `UI/` directory, our design system, the component should go under `Components/Core/IssueCard` for now.  
- `pullRequest` -> `issue`
- You have `url` but you are not using it, I'd expect that you can click `Artemis #12` and it opens to the issue link
- When I change `OPEN` to `CLOSED` nothing changes, it should change to the correct icon with the correct color, open does also not use the correct color
- We don't need `repository.nameWithOwner`, `repository.defaultBranch`, `repository.visibility`, `repository.url` in my opinion, just the name, then you also don't need to nest it
- `createdAt` could be a `dayjs` date or so that is correctly formatted in the UI
- There is the `fileDiff` icon but it is not in the github red for me, maybe there are some Storybook issues? Also I don't know how to configure it in the props
- There are two labels: `bug` and `enhancement`. I also can't see them in the input control object
- It would be nicer if you can configure each of the component's properties individually instead of the untyped nested object

FelixTJDietrich reviewed (changes requested):
Try running `npm run storybook` and play around with the story a bit if you can configure everything and everything looks right. We try to build dumb components in isolation and use storybook to document it. In the future it would be nice that a student can just click through it and see what components exist and also try out the different properties of the component.

FelixTJDietrich commented:
This is currently the component API:
```
<app-issue-card
  [pullRequest]="{title: 'Add feature X', number: 12, deletions: 5, url: 'http://example.com', state: 'CLOSED', repository: {name: 'Artemis', nameWithOwner: 'artemis-education/artemis', defaultBranch: 'master', visibility: 'PUBLIC', url: 'http://example.com'}, createdAt: 'Jan 1', pullRequestLabels: {'_constructor-name_':'Set'}, reviews: {'_constructor-name_':'Set'}}"
></app-issue-card>
```

Maybe we can not pass a complex type there, does it make sense, maybe not? Think like SwiftUI previews, this is basically what Storybook is for us

GODrums reviewed (commented):
Just merged the changes from develop into the branch so `npm run storybook` works again and I can give quick feedback :)

UI looks very nice already, would love to start using this in one of the next leaderboard features, such as the user pages. 
I can second @FelixTJDietrich's suggestion of having a clickable element to get to the Github PR page directly.

Do let me know if you need help with any elements of Felix's feedback, which aren't addressed yet!

iam-flo commented:
@GODrums I added a hover effect for the component, so that it is clearly clickable. I. think I implemented all requested changes. If you want to support, you can take a look at making the labels look like from GitHub. Haven't been able to implement that. In the future please use rebase for PRs as this cleans up the git tree.  I  will squash the commits before merge to clean everything up.

GODrums commented:
@iam-flo Sorry, I was out of line committing in your PR without asking. I think I also forgot to `git pull` before merging since I was doing it from mobile.
Our usual convention is that @FelixTJDietrich merges all PRs with squashing, so no need to worry about git trees.

I think there are few smaller QoL things of the feedback left, I'll have a quick look and mark them.

GODrums commented on webapp/src/app/core/IssueCard/issue-card.component.html:
```diff
@@ -0,0 +1,37 @@
+  </div>
+  <div class="flex gap-1 flex-wrap">
+    @for (label of pullRequestLabels(); track label.name) {
+      <span class="px-2 py-1 rounded-full text-xs mr-2 text-white" [style.background-color]="label.color">{{ label.name }}</span>
```
Github uses CSS's `calc()`, which I'm not familiar with either. But we could reduce the contrast with something like this:
```suggestion
      <span class="px-2 py-1 rounded-full text-xs mr-2 text-neutral-100/95" [style.background-color]="label.color">{{ label.name }}</span>
```

GODrums commented on webapp/src/app/core/IssueCard/issue-card.component.html:
```diff
@@ -0,0 +1,37 @@
+  </div>
+
+  <div class="flex justify-between font-bold text-sm mb-3">
+    {{ title() }}
```
If we remove the `(click)` and `cursor-pointer` from the main card `div` (not included in this suggestion) and move it to the title itself, we could create an interactive element like this, making it very similar to Github's own cards:
```suggestion
    <div class="hover:text-github-accent-foreground cursor-pointer" (click)="openIssue()">
      {{ title() }}
    </div>
```
Reference (on title hover):
![image](https://github.com/user-attachments/assets/7a88f8cb-27f2-48b9-bbf8-2245271fb720)

GODrums commented on webapp/src/app/core/IssueCard/issue-card.component.ts:
```diff
@@ -0,0 +1,41 @@
+  protected readonly octGitPullRequestClosed = octGitPullRequestClosed;
+
+  getMostRecentReview() {
+    return Array.from(this.reviews() || []).reduce((latest: PullRequestReview, review: PullRequestReview) => {
```
Since reviews is a required array, we don't really need the explicit conversion or `null`-check. Same for the types, which are inferred by Typescript:
```suggestion
    return this.reviews().reduce((latest, review) => {
```

GODrums commented on webapp/src/app/core/IssueCard/issue-card.component.ts:
```diff
@@ -0,0 +1,41 @@
+  repositoryName = input.required<string>();
+  reviews = input.required<Array<PullRequestReview>>();
+  createdAt = input.required<Dayjs>();
+  state = input.required<string>();
```
You can use the OpenAPI generated type here (make sure to import it):
```suggestion
  state = input.required<PullRequest.StateEnum>();
```

iam-flo commented:
@GODrums thank you for the suggestions 😄

FelixTJDietrich commented:
LGTM, Thanks :)
